File: backend/src/arxiv_digest/clients/email.py
"""Email client: send the weekly HTML digest over SMTP.

Async SMTP via ``aiosmtplib``. Host/port/TLS come from ``config.toml`` and credentials from
the environment, so the same client targets Gmail (the default) or any other provider without
code changes. The ``From`` address defaults to ``email_from``, falling back to the SMTP username.
"""

from email.message import EmailMessage

import aiosmtplib

from arxiv_digest.config import settings

_TEXT_FALLBACK = "This is the weekly arXiv ML digest. View it in an HTML-capable email client."


class EmailError(RuntimeError):
    """Raised when sending an email over SMTP fails."""


def _from_address() -> str:
    """The configured sender, falling back to the SMTP login when ``email_from`` is blank."""
    return settings.email_from or settings.smtp_username.get_secret_value()


async def send_email(*, to: str, subject: str, html: str) -> None:
    """Send one HTML email (with a plain-text fallback) to a single recipient."""
    message = EmailMessage()
    message["From"] = _from_address()
    message["To"] = to
    message["Subject"] = subject
    message.set_content(_TEXT_FALLBACK)
    message.add_alternative(html, subtype="html")
    try:
        await aiosmtplib.send(
            message,
            hostname=settings.smtp_host,
            port=settings.smtp_port,
            start_tls=settings.smtp_starttls,
            username=settings.smtp_username.get_secret_value(),
            password=settings.smtp_password.get_secret_value(),
        )
    except aiosmtplib.SMTPException as exc:
        msg = f"failed to send digest to {to}: {exc}"
        raise EmailError(msg) from exc

File: supabase/functions/_shared/email.ts
// Send mail over SMTP, reusing the same Gmail credentials as the weekly digest pipeline.
// Configured via function secrets: SMTP_USERNAME, SMTP_PASSWORD, and optionally SMTP_HOST /
// SMTP_PORT (default Gmail on 465/TLS). Low volume — one mail per signup or unsubscribe link.
import { SMTPClient } from 'https://deno.land/x/denomailer@1.6.0/mod.ts';

export async function sendEmail(to: string, subject: string, html: string): Promise<void> {
  const username = Deno.env.get('SMTP_USERNAME');
  const password = Deno.env.get('SMTP_PASSWORD');
  if (!username || !password) {
    throw new Error('SMTP_USERNAME and SMTP_PASSWORD must be set as function secrets');
  }
  const hostname = Deno.env.get('SMTP_HOST') ?? 'smtp.gmail.com';
  const port = Number(Deno.env.get('SMTP_PORT') ?? '465');

  const client = new SMTPClient({
    connection: {
      hostname,
      port,
      // Port 465 is implicit TLS; 587 upgrades via STARTTLS (tls: false lets denomailer do that).
      tls: port === 465,
      auth: { username, password },
    },
  });
  try {
    await client.send({ from: username, to, subject, content: 'text/html', html });
  } finally {
    await client.close();
  }
}
